### PR TITLE
Offline view requests

### PR DESCRIPTION
**Pull request by kieter**
This pull request covers US 08.02.01 #145 and US 08.01.01 #144 (View driver and rider requests while offline). 

To try this build then log in first while online, refresh by pulling down (or make a request) then close the app, go in airplane mode, and log in again.

~~I'll fix the merge conflicts soonnn.~~

### CODE REVIEW

File: app/src/main/java/comcmput301f16t01/github/carrier/FareCalculator.java [hunk 1]
```diff
@@ -18,6 +18,12 @@
     static final double BOOKING_FEE = 1.65;
     static final int MIN_FARE = (int) 5.00 * 100;
 
+    /**
+     * Estimates a fare between two locations given the distance and duration for them.
+     * @param distance Distance measured in kilometers
+     * @param duration Duration measured in seconds
+     * @return Returns an integer estimate
+     */
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/FareCalculator.java:36]: yaaassssss

File: app/src/main/java/comcmput301f16t01/github/carrier/MainActivity.java [hunk 2]
```diff
@@ -138,8 +141,17 @@ public void onPageScrollStateChanged(int state) {
...
         driver_fab.hide();
 
-        // Perform an update using RequestController
-        rc.performAsyncUpdate();
+        // Perform an update using RequestController if there is internet
+        if (ConnectionChecker.isThereInternet()) {
+            rc.performAsyncUpdate();
+        }
+        else {
+            Toast.makeText(this, "Offline", Toast.LENGTH_SHORT).show();
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/MainActivity.java]: can you change this toast to "You are offline."

File: app/src/main/java/comcmput301f16t01/github/carrier/MainActivity.java [hunk 3]
```diff
@@ -138,8 +141,17 @@ public void onPageScrollStateChanged(int state) {
...
+            rc.fetchAllRequestsWhereRider(UserController.getLoggedInUser());
+            rc.getOfferedRequests(UserController.getLoggedInUser());
+            //rc.saveRiderRequests();
+            Toast.makeText(this, "Size of requestsWhereRider: " + Integer.toString(rc.getRiderInstance().size()), Toast.LENGTH_SHORT).show();
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/MainActivity.java]: do we still need this?

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/ElasticRequestController.java [hunk 4]
```diff
@@ -525,13 +527,18 @@ private RequestList getRequests(List<Offer> offers) {
 
         @Override
         protected void onPostExecute(RequestList requests) {
-            // Perform result update on UI thread
-            if (withAsync) {
-                RequestController rc = new RequestController();
-                rc.getOffersInstance().replaceList( requests );
-                notifyListener();
+            // Perform result update on UI thread if there is internet
+            //TODO: ask bennett about this
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/ElasticRequestController.java]: did you ask bennnett? can you remove this todo? whats going on?

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 5]
```diff
@@ -211,13 +254,18 @@ public void searchByLocation(Location location) {
      */
     // TODO rename this method? i.e. getRequestsWhereDriverOffered, or something
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: are you going to rename it? I do not think I care to what but you should get rid of the TODO

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 6]
```diff
@@ -327,4 +389,83 @@ public void setRequestDescription(Request request, String description) {
...
+     */
+    public void saveRiderRequests() {
+        try {
+            FileOutputStream fos = saveContext.openFileOutput(this.RIDER_FILENAME, 0);
+            BufferedWriter out = new BufferedWriter(new OutputStreamWriter(fos));
+
+            Gson gson = new Gson();
+            gson.toJson(requestsWhereRider, out);
+            out.flush();
+            Log.i("Saving", "saveRiderRequests saved " + requestsWhereRider.size());
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: remove the log stuff

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 7]
```diff
@@ -327,4 +389,83 @@ public void setRequestDescription(Request request, String description) {
...
+
+    /**
+     * For offline functionality. Loads the cached riderrequests.
+     */
+    public void loadRiderRequests() {
+        FileInputStream fis = null;
+        try {
+            fis = saveContext.openFileInput(RIDER_FILENAME);
+            BufferedReader in = new BufferedReader(new InputStreamReader(fis));
+            // For debugging
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: delete this commented out stuff please?

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java (hunk 6 above)
Comment (MandyMeindersma) [reply app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: and error catching stuff?

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 8]
```diff
@@ -327,4 +389,83 @@ public void setRequestDescription(Request request, String description) {
...
+     * Caches the requests that the driver offered to fulfill.
+     */
+    public void saveDriverOfferedRequests() {
+        try {
+            FileOutputStream fos = saveContext.openFileOutput(this.DRIVER_FILENAME, 0);
+            BufferedWriter out = new BufferedWriter(new OutputStreamWriter(fos));
+
+            Gson gson = new Gson();
+            gson.toJson(requestsWhereOffered, out);
+            out.flush();
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:417]: delete debugging?

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginActivity.java [hunk 9]
```diff
@@ -21,6 +25,8 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_login);
 
+        //TODO: if online, quicklogin, else, offlinelogin
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginActivity.java]: why are these TODO's here?

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginActivity.java [hunk 10]
```diff
@@ -21,6 +25,8 @@ protected void onCreate(Bundle savedInstanceState) {
...
+
         tryQuickLogin();
 
         // TODO grab their username based on their ID?
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginActivity.java:32]: why this TODO here? have you solved it? just make them log in manually?

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java [hunk 11]
```diff
@@ -38,6 +47,23 @@ public void saveUsername( String username ) {
...
+            if (userToCache == null) {
+                out.write("");
+            } else {
+                Gson gson = new Gson();
+                gson.toJson(userToCache, out);
+            }
+            out.flush();
+            fos.close();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java:79]: what is e.printstackTrace?

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java [hunk 12]
```diff
@@ -54,4 +80,19 @@ public String loadUsername() {
...
+        FileInputStream fis = null;
+        User lastLoggedInUser = null;
+        try {
+            fis = saveContext.openFileInput(USER_FILENAME);
+            BufferedReader in = new BufferedReader(new InputStreamReader(fis));
+            Gson gson = new Gson();
+            Type type = new TypeToken<User>() {}.getType();
+            lastLoggedInUser = gson.fromJson(in, type);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java]: gah thing kieter

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/UserController.java [hunk 13]
```diff
@@ -115,6 +115,14 @@ public boolean logInUser(String username) {
         return true;
     }
 
+    public boolean offlineLogInUser(String usernameToLogin, User cachedUser) {
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/UserController.java]: Java docs? comments? anything?

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java [hunk 14]
```diff
@@ -54,4 +80,19 @@ public String loadUsername() {
 
         return username;
     }
+
+    public User loadUser() {
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java:109]: java docs comments? anything?

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java [hunk 15]
```diff
@@ -38,6 +47,23 @@ public void saveUsername( String username ) {
         }
     }
 
+    public void saveUser(User userToCache) {
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginMemory.java:65]: java docs comment annything

File: app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginActivity.java [hunk 16]
```diff
@@ -45,17 +51,26 @@ private void tryQuickLogin() {
...
 
-        if ( !uc.logInUser( username ) ) {
-            EditText usernameEditText = (EditText) findViewById( R.id.EditText_username );
-            usernameEditText.setText( username );
-            AlertDialog.Builder adb = new AlertDialog.Builder( this );
-            String message = "Your account '" + username + "' was not found!";
-            adb.setTitle( "Warning!" );
-            adb.setMessage( message );
-            adb.setPositiveButton( "OK", null );
+        if (ConnectionChecker.isThereInternet()) {
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Users/LoginActivity.java:57]: This is about line 37. could your java doc comments be more descriptive? more than one line if possible?

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 17]
```diff
@@ -327,4 +389,83 @@ public void setRequestDescription(Request request, String description) {
...
+    public void loadDriverOfferedRequests() {
+        FileInputStream fis = null;
+        try {
+            fis = saveContext.openFileInput(DRIVER_FILENAME);
+            BufferedReader in = new BufferedReader(new InputStreamReader(fis));
+            Gson gson = new Gson();
+            Type listType = new TypeToken<RequestList>() {}.getType();
+            requestsWhereOffered.replaceList((RequestList) gson.fromJson(in, listType));
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:437]: ew. stack trace

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 18]
```diff
@@ -327,4 +389,83 @@ public void setRequestDescription(Request request, String description) {
...
+            Log.i("Saving", "saveRiderRequests saved " + requestsWhereRider.size());
+            fos.close();
+
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * For offline functionality. Loads the cached riderrequests.
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: change to "rider requests"

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 19]
```diff
@@ -256,10 +310,16 @@ public RequestList fetchRequestsWhereRider(User rider, Integer... statuses ) {
             e.printStackTrace();
         }
         requestsWhereRider.replaceList( foundRequests );
+        saveRiderRequests();
         return foundRequests;
     }
 
     public RequestList fetchAllRequestsWhereRider( User rider ) {
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: java doc?

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 20]
```diff
@@ -237,6 +285,12 @@ public void clearAllRiderRequests(User rider) {
      * @return A list of requests from the given criteria
      */
     public RequestList fetchRequestsWhereRider(User rider, Integer... statuses ) {
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: java doc

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 21]
```diff
@@ -211,13 +254,18 @@ public void searchByLocation(Location location) {
      */
     // TODO rename this method? i.e. getRequestsWhereDriverOffered, or something
     public RequestList getOfferedRequests(User driver) {
+        if (!ConnectionChecker.isThereInternet()) {
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:248]: java doc for this method

File: app/src/main/java/comcmput301f16t01/github/carrier/Notifications/ConnectionChecker.java [hunk 22]
```diff
@@ -24,4 +35,48 @@ public static boolean isConnected( Context context ) {
...
+        protected Boolean doInBackground(Void... params) {
+            URL team1URL = null;
+            // Try making an http request to our index on elastic search
+            try {
+                team1URL = new URL("http://cmput301.softwareprocess.es:8080/cmput301f16t01");
+                HttpURLConnection urlConnection = (HttpURLConnection) team1URL.openConnection();
+                // Websites return status code 200 (HTTP_OK in this API) when you've successfully connected to them
+                return urlConnection.getResponseCode() == HttpURLConnection.HTTP_OK;
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Notifications/ConnectionChecker.java:60]: get rid of the stack thing

File: app/src/main/java/comcmput301f16t01/github/carrier/Notifications/ConnectionChecker.java [hunk 23]
```diff
@@ -24,4 +35,48 @@ public static boolean isConnected( Context context ) {
                 activeNetwork.isConnectedOrConnecting();
     }
 
+    //TODO: http://stackoverflow.com/questions/15496278/httpurlconnection-is-throwing-exception
+    //TODO: https://developer.android.com/reference/java/net/HttpURLConnection.html
```
Comment (MandyMeindersma) [inline app/src/main/java/comcmput301f16t01/github/carrier/Notifications/ConnectionChecker.java:38]: did you tell sarah?

**Review by MandyMeindersma — CHANGES_REQUESTED** [review]
I know this is tedious but could you go through each comment and add a 'done' once you have done it...

File: app/src/main/java/comcmput301f16t01/github/carrier/Notifications/ConnectionChecker.java (hunk 23 above)
Comment (kieter) [reply app/src/main/java/comcmput301f16t01/github/carrier/Notifications/ConnectionChecker.java:38]: I told sarah

**Comment by kieter** [comment]
DONE. I wrote some tests tooo @MandyMeindersma

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java (hunk 20 above)
Comment (Hreherch) [reply app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java]: It's in my branch.

**Comment by MandyMeindersma** [comment]
lols more conflicts

**Comment by MandyMeindersma** [comment]
@kieter am I safe to assume you have done everything I have commented?

**Comment by sehoven** [comment]
I'm forcing this because Mandy was good with everything except for merge conflicts and we want to merge this branch before any others get merged first accidentally and force Kieter to fix more merge conflicts.

**Pull request merged**